Wp-env: Add phpunit support

noahtallen opened the pull request:
This PR allows the phpunit script to work with @wordpress/env.

## Changes:
- Add phpunit service to wp-env docker
- Add wp_phpunit composer dependency so we don't have to use wordpress-develop
- switch package.json to use wp-env for phpunit tests.

To test, run:

```sh
# Get the new composer changes.
./packages/env/bin/wp-env run composer 'composer install'  

# Run phpunit via package.json:
npm run test-unit-php

# With multisite setup:
npm run test-unit-php-multisite

# Alternatively, directly run phpunit command (the package commands are aliases for this)
./packages/env/bin/wp-env run phpunit 'phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist'   
```

## External API

This works pretty well for us internally, but I think it will be difficult to use from a 3rd party perspective for a few reasons:

1. Currently, you are required to know the absolute path to the phpunit config file inside Docker. That will be different for every source.
2. You have to install the correct wp-phpunit dependency yourself.
3. You have to know how to setup the bootstrap and config files correctly. 

See #22365 for explorations on how to improve the external API.
 
### To do:
- [x] Should we export functions from @wordpress/env to handle things like `getDockerComposeDir`? (or even for reading the config file if needed.)
- [x] Provide a phpunit service in the default docker config. ([Similar to the WordPress one we used before](https://github.com/WordPress/wordpress-develop/blob/112649ea924e2aed6c18bfcbf8ab7306d3c53b78/docker-compose.yml#L95-L121).)
- [x] verify multisite tests also work
- [x] verify exit code is 1 if tests fail
- [x] use different tests port in travis to avoid conflicting with scripts/env
- [x] verify php version from `LOCAL_PHP` is used in phpunit environment
- [x] make sure wordpress is installed before continuing
- [ ] make sure we have permissions to rename files in travis

epiqueras reviewed (changes requested):
>Should we export functions from @wordpress/env to handle things like getDockerComposeDir? (or even for reading the config file if needed.)

The `wp-env run` command removes the need for any of that.

 >Provide a phpunit service in the default docker config. (Similar to the WordPress one we used before.)

Sure

 >In order to remove the wp-env package.json field, what if we added a primaryPluginDir field to the .wp-env.json file? If not, we'll need to pass it through args and/or default to cwd.

It should be `cwd` like with other scripts.

noahtallen commented:
Thanks, @epiqueras! I've modified this to call `wpEnv.run` instead (which adds a dependency on @wordpress/env.)

Though this raises the question: test-php is needed anymore?

In a project, I can add: `"test:php": "wp-env run phpunit phpunit -c /var/www/src/wp-content/plugins/full-site-editing-plugin/phpunit.xml.dist",` to my `package.json` file and it should work so long as we have a phpunit service in the default docker config (which I'll look at adding next)

epiqueras commented:
I guess we need them for backwards compatibility.

I don’t think importing wp-env is a good idea. Why not make it a peer
dependency and invoke the bin file like the other scripts do.

noahtallen committed:
Create phpunit service in docker config

- install wp-phpunit via composer
- include composer autoload and config in bootstrap
- add missing config options
- allow service to be run without starting wp-env first

noahtallen commented:
> so long as we have a phpunit service in the default docker config

I've added the basic configuration for the phpunit service but ran into an issue. Previously, we always used the wordpress-develop repository, which includes the WordPress test files all the required bootstrapping. Since we are now using the production image, those directories don't exist.

This means that WordPress phpunit bootstrap scripts will fail (like this):

https://github.com/WordPress/gutenberg/blob/4030f736ddfc34ecbfe254ed65f571a31b272a67/phpunit/bootstrap.php#L28-L31

It's possible we'll have to add a requirement to use `WordPress/wordpress-develop` in order to run phpunit. Otherwise, we'll have to make sure those test files are downloaded to the Docker image somehow.

epiqueras commented:
Ideally, the test bootstrap code would be a separate composer package that we load in isolation. I've heard that was planned for Core, but it's not a priority. The issue with adding the requirement to use `wordpress-develop` is that we force ourselves to build it and lose the excellent speed `wp-env start` has right now.

epiqueras commented:
Maybe we should require it only for that test script.

gziolo commented:
@noahtallen - nice to see progress on migrating PHPUnit tests to `wp-env` 💯 

Is it necessary to update `wp-scripts env` to make it work? It feels like we are going to deprecate the whole `env` subset of scripts and advice to use `wp-env` instead. Once everything is working with `wp-env` on Travis, there is no reason to keep two competing implementations.

epiqueras commented:
Agreed, I'd rather see this PR deprecating `wp-scripts env` and replacing its usage with `wp-env` everywhere instead of doing it in a follow-up. It will help catch any possible bugs.

noahtallen commented:
Definitely agree! How would we go about deprecating those scripts? Is there a previous example I could look at for how we typically handle these things?

(Or do we just delete those files and the references to them and move on?)

> replacing its usage with wp-env everywhere

I think we'll only need to update phpcs and phpunit within the scripts package, which is convenient, but as we can see here it might not work out of the box :)

epiqueras commented:
Use the `deprecate` utility to add messages, add a change-log entry, and write a dev note.

noisysocks commented:
> Ideally, the test bootstrap code would be a separate composer package that we load in isolation. I've heard that was planned for Core, but it's not a priority. The issue with adding the requirement to use `wordpress-develop` is that we force ourselves to build it and lose the excellent speed `wp-env start` has right now.

I think it's probably worth biting the bullet and doing this. A WordPress plugin that wants to use `WP_UnitTestCase` shouldn't have to vendor all of `wordpress-develop` in order to do so.

epiqueras commented:
Agreed, who would be the best person to work on that?

noisysocks commented:
My DMs are open if somebody is interested in doing this and needs help navigating Trac or the codebase 😄

noahtallen commented:
> My DMs are open if somebody is interested in doing this and needs help navigating Trac or the codebase

Nice :) I'm afraid my cycles got taken up by a different project at the moment, so I probably won't be able to revisit this PR for a while. (Maybe days, more likely weeks)

talldan commented:
I haven't been able to find any reference in trac to making the unit test bootstrap a separate composer package. Would you be able to share that discussion if you can find it @epiqueras, @noisysocks?

noahtallen commented:
Also, I'd be interested in poking around at that, but I'm not sure how much work it would take, or even what steps need to be taken to make it happen

noisysocks commented:
I also can't find any discussion about this, but I spoke with @johnbillion who suggested that Gutenberg could use the third party `wp-phpunit/wp-phpunit` package, and to look at this PR for inspiration on what needs to happen: https://github.com/johnbillion/user-switching/pull/38

chrisvanpatten commented:
I would LOVE this to happen. Setting up the core unit test bootstrap is such a pain — smoothing over that with `wp-env` would be incredible. Are there any action items that could help move this forward?

ockham commented:
Rebased.

ockham commented:
> I also can't find any discussion about this, but I spoke with @johnbillion who suggested that Gutenberg could use the third party `wp-phpunit/wp-phpunit` package, and to look at this PR for inspiration on what needs to happen: [johnbillion/user-switching#38](https://github.com/johnbillion/user-switching/pull/38)

Thanks! This looks promising. If we're lucky, we 'just' need to follow [these steps](https://github.com/wp-phpunit/docs); tl;dr:

- `composer require wp-phpunit/wp-phpunit --dev`
- Add a tests WP config file (unless there is one already in Gutenberg?)
- Point our tests framework to wp-phpunit, to set the `'WP_PHPUNIT__TESTS_CONFIG` env var -- probably in `phpunit.dist.xml` ([see](https://github.com/wp-phpunit/docs#configuration)). Or maybe we have to set `WP_TESTS_DIR` for use in our [bootstrap](https://github.com/WordPress/gutenberg/blob/90d3e6cbe0ffb38babe8aa66e21e81805ed148cf/phpunit/bootstrap.php#L15) (which is what's [failing](https://github.com/WordPress/gutenberg/pull/20090#issuecomment-583691793) now as pointed out by @noahtallen).

I'm calling it a day soon and won't be working tomorrow, but I'm documenting this here in case someone else wants to pick it up and play with it. Otherwise I'll get back to it next week :smile:

ockham commented:
Update: I tried [the following](https://github.com/wp-phpunit/docs#examples) today:

```diff
diff --git a/phpunit/bootstrap.php b/phpunit/bootstrap.php
index 17989ee336..6cb84ff456 100644
--- a/phpunit/bootstrap.php
+++ b/phpunit/bootstrap.php
@@ -5,6 +5,8 @@
  * @package Gutenberg
  */
 
+require_once dirname( dirname( __FILE__ ) ) . '/vendor/autoload.php';
+
 // If we're running in WP's build directory, ensure that WP knows that, too.
 if ( 'build' === getenv( 'LOCAL_DIR' ) ) {
 	define( 'WP_RUN_CORE_TESTS', true );
@@ -12,7 +14,7 @@ if ( 'build' === getenv( 'LOCAL_DIR' ) ) {
 
 // Determine the tests directory (from a WP dev checkout).
 // Try the WP_TESTS_DIR environment variable first.
-$_tests_dir = getenv( 'WP_TESTS_DIR' );
+$_tests_dir = getenv( 'WP_TESTS_DIR' ) ?: getenv( 'WP_PHPUNIT__DIR' );
 
 // See if we're installed inside an existing WP dev instance.
 if ( ! $_tests_dir ) {
```

Running `npx wp-env run phpunit "phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist"` then gave an error complaining that `WP_TESTS_DOMAIN` and `ABSPATH` were undefined. 

As for setting `WP_TESTS_DOMAIN`, I found an [option](https://github.com/docker-library/wordpress/pull/142) for the `wordpress` Docker image that allows to inject constants like this into `wp-config.php`, so I tried

```diff
diff --git a/packages/env/lib/build-docker-compose-config.js b/packages/env/lib/build-docker-compose-config.js
index dc2c43794e..18ea547c26 100644
--- a/packages/env/lib/build-docker-compose-config.js
+++ b/packages/env/lib/build-docker-compose-config.js
@@ -159,6 +159,8 @@ module.exports = function buildDockerComposeConfig( config ) {
                                ],
                                environment: {
                                        LOCAL_DIR: 'html',
+                                       WORDPRESS_CONFIG_EXTRA:
+                                               "define( 'WP_TESTS_DOMAIN', 'example.org' );\n",
                                },
                        },
                },
```

While this generated the correct `docker-compose.yml`,

```yaml
    environment:
      LOCAL_DIR: html
      WORDPRESS_CONFIG_EXTRA: |
        define( 'WP_TESTS_DOMAIN', 'example.org' );
```

it had no impact on the `wp-config.php` :slightly_frowning_face: I also tried the same with an 'actual' `wordpress` image, in the `tests-wordpress` service definiton:

```diff
diff --git a/packages/env/lib/build-docker-compose-config.js b/packages/env/lib/build-docker-compose-config.js
index dc2c43794e..9835b55fd3 100644
--- a/packages/env/lib/build-docker-compose-config.js
+++ b/packages/env/lib/build-docker-compose-config.js
@@ -130,6 +130,8 @@ module.exports = function buildDockerComposeConfig( config ) {
                                ports: [ testsPorts ],
                                environment: {
                                        WORDPRESS_DB_NAME: 'tests-wordpress',
+                                       WORDPRESS_CONFIG_EXTRA:
+                                               "define( 'WP_TESTS_DOMAIN', 'example.org' );\n",
                                },
                                volumes: testsMounts,
                        },
```

Once again, the generated `docker-compose.yml` looked fine, but the `wp-config.php` didn't change :slightly_frowning_face: 

`ABSPATH` OTOH is defined in `wp-config.php`, meaning that that config file isn't being picked up anyway tho :slightly_frowning_face:

noahtallen commented:
> As for setting WP_TESTS_DOMAIN, I found an option for the wordpress Docker image that allows to inject constants like this into wp-config.php, so I tried

I think we have a way to do this with wp-env as well, and I wonder if it is overwriting what we add to the docker config here: https://github.com/WordPress/gutenberg/tree/master/packages/env#wp-envjson

In `config.js`, you could try adding default values for those similar to the existing `config: { WP_DEBUG: true, SCRIPT_DEBUG: true }`

noahtallen commented:
I think I got it working. I:

1. Added `WP_PHPUNIT__TESTS_CONFIG` as an environment variable in the docker config. PHP picks that up so wp_phpunit has access to it.
2. Added the missing PHP constants to the default config options. (In the future, we can add these keys to a specific test config, but that's not yet supported.)

Example from running it in Gutenberg:

```
~/source/gutenberg(try/phpunit-tests-with-wp-env*) » ./packages/env/bin/wp-env run phpunit "phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist"                                                                                          noah.allen@Noahs-MacBook-Pro
⠙ Running `phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist` in 'phpunit'.


Notice: Constant WP_LANG_DIR already defined in /var/www/html/wp-content/plugins/gutenberg/vendor/wp-phpunit/wp-phpunit/includes/bootstrap.php on line 60

Notice: Constant WP_MEMORY_LIMIT already defined in /var/www/html/wp-content/plugins/gutenberg/vendor/wp-phpunit/wp-phpunit/includes/bootstrap.php on line 72

Notice: Constant WP_MAX_MEMORY_LIMIT already defined in /var/www/html/wp-content/plugins/gutenberg/vendor/wp-phpunit/wp-phpunit/includes/bootstrap.php on line 73
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

...............................................................  63 / 101 ( 62%)
......................................                          101 / 101 (100%)

Time: 3.8 seconds, Memory: 40.50 MB

OK (101 tests, 1801 assertions)



✔ Ran `phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist` in 'phpunit'. (in 4s 989ms)
```

As you can see, there are some redeclared constants. Not sure what to do about those 🤔

noahtallen commented:
`npm run test-unit-php` also works correctly now assuming a composer install has occurred.

Outstanding items:
- [ ] Should we move `test-php.js` outside of `scripts/env` alongside the other scripts, since scripts/env is being deprecated? We could also ditch it entirely in favor of calling wp-env from a package.json run command.
- [ ] Is there anything to do about the re-declared constants? They happen inside the wp-phpunit dependency, so I'm not sure.
- [ ] `npm run test-php` does not work for some reason. Something to do with the lint check doesn't seem to be right 🔍

chrisvanpatten commented on packages/scripts/scripts/env/test-php.js:
```diff
@@ -1,21 +1,39 @@
+	.run( {
+		container: 'phpunit',
+		command: [
+			`phpunit -c /var/www/${ localDir }/wp-content/plugins/${ pluginDir }/phpunit.xml.dist ${ args }`,
```
Would providing your own `-c` flag override the default (e.g. does last in get precedence?) Would be good if this could be easily overridden.

noahtallen replied:
Hm, that's a good idea. I'm not sure if it gets overridden. I'm not sure how to best expose this because it needs to know the path to the config file _inside_ docker, and then join that with the path to the config in the actual source code. 

So it's also currently bad because it won't work for wordpress or theme sources.

noahtallen replied:
I've added this to wp-env directly, so you could set the path via .wp-env.json. (it uses `$cwd/phpunit.dist.xml` by default if it exists)

noahtallen commented:
> Something to do with the lint check doesn't seem to be right

see investigation here: https://github.com/WordPress/gutenberg/pull/21537#issuecomment-627625944. tldr: composer output is considered an error even when it isn't one. It shouldn't block this PR though.

> Should we move test-php.js outside of scripts/env alongside the other scripts

We decided to remove the changes in test-php.js entirely since it is a scripts/env thing and not a wp-env thing. So we can keep that in place until we are ready to deprecate it entirely.

And then finally, to make running phpunit a lot easier, I added a `phpunit` command to wp-env. This basically calls `wp-env run phpunit phpunit -c $phpunit_config`, but does some work to figure out what `$phpunit_config` actually is inside the docker service. Otherwise, using phpunit would require knowledge of the internal wp-env workings, and I don't think we want that! This way, we can automatically use a phpunit config files located in, and override that with a specific one passed by .wp-env.json.

BTW, I also decided to _not_ start calling wp-env via `npm run test-unit-php`. I think once this PR is in, we can start working with #22303 to switch travis over to wp-env where we want.

noahtallen commented:
after some discussion with @epiqueras, I removed the phpunit command and instead added cwd support to `wp-env run`. So you can run phpunit with:

`wp-env run phpunit phpunit ./`

1. first phpunit is the service name
2. seccond phpunit is the name of the command
3. The given path should contain a `phpunit.dist.xml` file. Though specified relative to cwd, it is converted internally to match the path to its parent source in the docker container. So if you put your config file at `gutenberg/phpunit/phpunit.dist.xml`, passing `./phpunit` will match the file in Docker. so it gets converted internally to `var/www/html/wp-content/plugins/gutenberg/phpunit` (this code is flexible enough to work with any mapped source)

noisysocks commented:
Can't `wp-env` use the actual `$CWD` variable? In other words, why `wp-env run phpunit phpunit ./some-directory` and not `cd ./some-directory; wp-env run phpunit phpunit`?

If possible I'd like to avoid adding any further positional arguments to `wp-env run` so that we can eventually modify `wp-env run <service> <command>` to treat `<command>` as a raw string containing spaces. This will make it easier to run things like `wp-env run cli wp post list`.

noahtallen commented:
> Can't wp-env use the actual $CWD variable?

two thoughts:
- if it uses $CWD directly, does it use that for wp-env run all the time? (I.e. cd to “wp-content/plugins/gutenberg before every run command) And then how would you override it without passing some kind of option?
- if you added it as part of the command you pass in instead of having it baked in to every command, then you’d have to know the internal layout of the docket filesystem for it to be a valid path. 

Also just to make sure we’re on the same page, the main difference between those variants is that we need to accept a path for the local filesystem (e.g. “./” as in, my local Gutenberg git checkout from which I run wp-env), and give to docker a path that works in docker (e.g. “var/www/html/wp-content/plugins/gutenberg”).

The main problem I want to solve in this PR is: let the user easily run phpunit without needing knowledge of how wp-env works under the hood. I can run “wp-env run phpunit ‘phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.dist.xml” and it works without any of the boilerplate. But it requires me to know how the plugin was mapped internally and I don’t think that’s good.

Having a separate phpunit command for wp-env solves the problem in a very smooth way for the user. Enrique’s suggestion to just add cwd support to the run command also solves the problem. Admittedly, it’s not as “clean” an API, but it works without having to add a new command which is basically just run with some boilerplate.

I’m open for suggestions on how to improve the API. If we want to keep discussing that, I suggest we merge basic phpunit support for the time being so that we can at least use it by passing the internal docker path to the config file. I know there is some php work being done right now that would really benefit from some unit tests making it in :)

noahtallen commented:
this PR now only exposes the phpunit and composer changes itself. it doesn't try to expose it in a nice way for other consumers. see #22365 for that. 

To test, run:

```sh
# Get the new composer changes.
./packages/env/bin/wp-env run composer 'composer install'  

# Run phpunit command pointing at Gutenberg config file.
./packages/env/bin/wp-env run phpunit 'phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist'   
```

noahtallen committed:
Add phpunit to run command

noahtallen commented:
currently checking two things:
1. multisite tests
2. make sure that exit code is not 0 if a test fails

epiqueras commented on packages/env/lib/build-docker-compose-config.js:
```diff
@@ -142,11 +142,24 @@ module.exports = function buildDockerComposeConfig( config ) {
 			},
+			phpunit: {
+				image: 'wordpressdevelop/phpunit',
+				depends_on: [ 'wordpress' ],
```
```suggestion
				depends_on: [ 'tests-wordpress' ],
```

Same for `tests-cli`.

noahtallen replied:
great idea

noahtallen committed:
Use tests-wordpress for service image

noahtallen committed:
Switch multisite tests over as well

noahtallen commented:
currently troubleshooting this error in CI. it looks like running the phpunit command fails because it doesn't like that tests-wordpress is already running, or something like that.

```
36.43s$ npm run test-php && npm run test-unit-php-multisite

> gutenberg@8.1.0 test-php /home/travis/build/WordPress/gutenberg

> npm run lint-php && npm run test-unit-php

> gutenberg@8.1.0 prelint-php /home/travis/build/WordPress/gutenberg

> npm run wp-env run composer install -- --no-interaction

> gutenberg@8.1.0 wp-env /home/travis/build/WordPress/gutenberg

> packages/env/bin/wp-env "run" "composer" "install" "--no-interaction"

- 

latest: Pulling from library/composer

Digest: sha256:7d49a2b42bf742f39e0f8e823612ab0def8a479379d9610275c28c39d746fde6

Status: Downloaded newer image for composer:latest

PHP CodeSniffer Config installed_paths set to ../../phpcompatibility/php-compatibility,../../sirbrillig/phpcs-variable-analysis,../../wp-coding-standards/wpcs

✔ Ran `install` in 'composer'. (in 11s 417ms)

> gutenberg@8.1.0 lint-php /home/travis/build/WordPress/gutenberg

> npm run wp-env run composer run-script lint

> gutenberg@8.1.0 wp-env /home/travis/build/WordPress/gutenberg

> packages/env/bin/wp-env "run" "composer" "run-script" "lint"

- 

............................................................  60 / 114 (53%)

......................................................       114 / 114 (100%)

Time: 2.89 secs; Memory: 30MB

✔ Ran `run-script lint` in 'composer'. (in 4s 695ms)

> gutenberg@8.1.0 test-unit-php /home/travis/build/WordPress/gutenberg

> wp-env run phpunit 'phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist'

- 

✖ Error while running docker-compose command.

latest: Pulling from library/mariadb

Digest: sha256:d96a3364fc783c02d19c54aa81e0dcf81a339340f74a4b3a64db1e72a419f5ab

Status: Downloaded newer image for mariadb:latest

latest: Pulling from library/wordpress

Digest: sha256:19faa564fd56e817cb6057ebacd3fda84fc2b4805a477aaf9eaa495c712f9e29

Status: Downloaded newer image for wordpress:latest

Pulling mysql (mariadb:)...

Pulling tests-wordpress (wordpress:)...

Creating ce047cf4da04af20491927334fb977c6_mysql_1 ... 

ERROR: for ce047cf4da04af20491927334fb977c6_tests-wordpress_1  Cannot start service tests-wordpress: driver failed programming external connectivity on endpoint ce047cf4da04af20491927334fb977c6_tests-wordpress_1 (39fa3b493882f2ce74659df096654f2ac142838d07ba071e17d35008c329b8fa): Bind for 0.0.0.0:8889 failed: port is already allocated

ERROR: for tests-wordpress  Cannot start service tests-wordpress: driver failed programming external connectivity on endpoint ce047cf4da04af20491927334fb977c6_tests-wordpress_1 (39fa3b493882f2ce74659df096654f2ac142838d07ba071e17d35008c329b8fa): Bind for 0.0.0.0:8889 failed: port is already allocated

Encountered errors while bringing up the project.
```

noahtallen commented:
Ahh. That's probably because the port is already allocated to the scripts/env instance 😆

noahtallen committed:
Update tests port number for travis

noahtallen commented on .travis.yml:
```diff
@@ -159,11 +159,12 @@ jobs:
         - npm run test-php && npm run test-unit-php-multisite
 
     - name: PHP unit tests (PHP 5.6)
-      env: LOCAL_PHP=5.6-fpm
```
We don't support changing the php version yet. Let's see how much it'll take 👀

noahtallen replied:
this ended up being pretty easy. :)

noahtallen committed:
Add php version support to phpunit command

noahtallen commented:
Hm. Now getting this error. The weird thing is that it does not cause phpunit to fail. I wonder if this could be caused by not having run the wordpress install yet 🤔 

```
> wp-env run phpunit 'phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist'

- 

latest: Pulling from library/mariadb

Digest: sha256:d96a3364fc783c02d19c54aa81e0dcf81a339340f74a4b3a64db1e72a419f5ab

Status: Downloaded newer image for mariadb:latest

latest: Pulling from library/wordpress

Digest: sha256:19faa564fd56e817cb6057ebacd3fda84fc2b4805a477aaf9eaa495c712f9e29

Status: Downloaded newer image for wordpress:latest

latest: Pulling from wordpressdevelop/phpunit

Digest: sha256:df71f7489bf95c4aa19e6de1ff581929f17d15c8a6d8a23a98310803cced0820

Status: Downloaded newer image for wordpressdevelop/phpunit:latest

<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" dir='ltr'>

<head>

	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />

	<meta name="viewport" content="width=device-width">

			<title>Database Error</title>

	<style type="text/css">

		html {

			background: #f1f1f1;

		}

		body {

			background: #fff;

			color: #444;

			font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif;

			margin: 2em auto;

			padding: 1em 2em;

			max-width: 700px;

			-webkit-box-shadow: 0 1px 3px rgba(0, 0, 0, 0.13);

			box-shadow: 0 1px 3px rgba(0, 0, 0, 0.13);

		}

		h1 {

			border-bottom: 1px solid #dadada;

			clear: both;

			color: #666;

			font-size: 24px;

			margin: 30px 0 0 0;

			padding: 0;

			padding-bottom: 7px;

		}

		#error-page {

			margin-top: 50px;

		}

		#error-page p,

		#error-page .wp-die-message {

			font-size: 14px;

			line-height: 1.5;

			margin: 25px 0 20px;

		}

		#error-page code {

			font-family: Consolas, Monaco, monospace;

		}

		ul li {

			margin-bottom: 10px;

			font-size: 14px ;

		}

		a {

			color: #0073aa;

		}

		a:hover,

		a:active {

			color: #00a0d2;

		}

		a:focus {

			color: #124964;

			-webkit-box-shadow:

				0 0 0 1px #5b9dd9,

				0 0 2px 1px rgba(30, 140, 190, 0.8);

			box-shadow:

				0 0 0 1px #5b9dd9,

				0 0 2px 1px rgba(30, 140, 190, 0.8);

			outline: none;

		}

		.button {

			background: #f7f7f7;

			border: 1px solid #ccc;

			color: #555;

			display: inline-block;

			text-decoration: none;

			font-size: 13px;

			line-height: 2;

			height: 28px;

			margin: 0;

			padding: 0 10px 1px;

			cursor: pointer;

			-webkit-border-radius: 3px;

			-webkit-appearance: none;

			border-radius: 3px;

			white-space: nowrap;

			-webkit-box-sizing: border-box;

			-moz-box-sizing:    border-box;

			box-sizing:         border-box;

			-webkit-box-shadow: 0 1px 0 #ccc;

			box-shadow: 0 1px 0 #ccc;

			vertical-align: top;

		}

		.button.button-large {

			height: 30px;

			line-height: 2.15384615;

			padding: 0 12px 2px;

		}

		.button:hover,

		.button:focus {

			background: #fafafa;

			border-color: #999;

			color: #23282d;

		}

		.button:focus {

			border-color: #5b9dd9;

			-webkit-box-shadow: 0 0 3px rgba(0, 115, 170, 0.8);

			box-shadow: 0 0 3px rgba(0, 115, 170, 0.8);

			outline: none;

		}

		.button:active {

			background: #eee;

			border-color: #999;

			-webkit-box-shadow: inset 0 2px 5px -3px rgba(0, 0, 0, 0.5);

			box-shadow: inset 0 2px 5px -3px rgba(0, 0, 0, 0.5);

		}

			</style>

</head>

<body id="error-page">

	<div class="wp-die-message"><h1>Error establishing a database connection</h1></div></body>

</html>

	

✔ Ran `phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist` in 'phpunit'. (in 21s 312ms)
```

noahtallen committed:
Try starting wp-env first.

noahtallen committed:
Set php binary once more

noahtallen commented:
Alright, I think we're down to the last error.

```
1) Extend_Styles_Test::test_without_built_styles

rename(/var/www/html/wp-content/plugins/gutenberg/build/editor/editor-styles.css,/var/www/html/wp-content/plugins/gutenberg/build/editor/editor-styles.css.bak): Permission denied
```

There's a php test that renames some CSS files and we don't have permission to do that in travis with wp-env apparently

noahtallen committed:
try chmod on build

noahtallen commented:
phpunit tests look good in Travis now. just some intermittent e2e failures now, so rerunning those.

ockham commented on packages/env/lib/config.js:
```diff
@@ -128,7 +128,12 @@ module.exports = {
+				config: {
+					WP_DEBUG: true,
+					SCRIPT_DEBUG: true,
+					WP_TESTS_DOMAIN: 'example.org',
+					WP_PHP_BINARY: 'php',
```
I'm not sure this is working :thinking: -- at least not for me :slightly_frowning_face: 

This is the `readConfig` function, which is used by `init-config.js` to write the `docker-compose.yml` file (via `build-docker-compose-config.js`). But AFAICS, none of those files evaluate the contents of the `config`  object. And if I check my `~/wp-env/12345678/docker-compose.yml`, I don't see any of those config constants either. :thinking:

noahtallen replied:
Right, the constants are not added to the docker compose config, but set with wp-cli:
https://github.com/WordPress/gutenberg/blob/066d8390989c40fe2f6e8fd79adc4c1546a28165/packages/env/lib/wordpress.js#L95-L112

This function is called during `wp-env start`
